File: 121-best-time-to-buy-and-sell-stock/solution.py
# Runtime: 771ms, Beats 65.25% of users with Python3
# Memory: 27.40MB, Beats 83.73% of users with Python3

"""
    -Let's say we're buying on the first day
    -We will iterate through the array to find the maximum profit:
        -If the current price is less than the price we bought at, we will buy at the current price
        -Otherwise, we will sell at the current price and update profit
"""
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxi=0
        buy=prices[0]
        for i in range(1,len(prices)):
            if(buy>prices[i]):
                buy=prices[i]
            else:
                maxi=max(maxi,prices[i]-buy)
        return maxi
        
        

File: 121-best-time-to-buy-and-sell-stock/solution.cpp
// Runtime: 86ms, Beats 84.72% of users with C++
// Memory: 95.69MB, Beats 23.82% of users with C++
//Brute Force will not be efficient and will get you a TLE

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int p=0,q=1,maxi=0,n=prices.size();
        while(q<n)
        {
            if(prices[q]>prices[p]) maxi=max(maxi,prices[q]-prices[p]);
            else p=q;
            q++;
        }
        return maxi;
    }
};